File: BalanceInfantry_Course/Simulation/leg_wbr.m
function [T1 T2] = ykb_caLc(F0,Tp,q1,q2)

% 赋值给定变量
L1a = 0.075;
L1u = 0.15;
L1d = 0.25;
L2a = 0.075;
L2u = 0.15;
L2d = 0.25;

%中间变量计算
X1 = L1a - L1u * cos(q1);
X2 = L2a - L2u * cos(q2);
Y1 = L1u * sin(q1);
Y2 = L2u * sin(q2);
sigma1 = sqrt(-(-L1d^2-2*L1d*L2d-L2d^2+X1^2+2*X1*X2+X2^2+Y1^2-2*Y1*Y2+Y2^2) * (-L1d^2+2*L1d*L2d-L2d^2+X1^2+2*X1*X2+X2^2+Y1^2-2*Y1*Y2+Y2^2));
Xe = (Y1*sigma1-Y2*sigma1+L1d^2*X1+L1d^2*X2-L2d^2*X1-L2d^2*X2+X1*X2^2-X1^2*X2-X1*Y1^2-X1*Y2^2+X2*Y1^2+X2*Y2^2-X1^3+X2^3+2*X1*Y1*Y2-2*X2*Y1*Y2) / (2*(X1^2+2*X1*X2+X2^2+Y1^2-2*Y1*Y2+Y2^2));
Ye = (X1*sigma1+X2*sigma1-L1d^2*Y1+L1d^2*Y2+L2d^2*Y1-L2d^2*Y2+X1^2*Y1+X1^2*Y2+X2^2*Y1+X2^2*Y2-Y1*Y2^2-Y1^2*Y2+Y1^3+Y2^3+2*X1*X2*Y1+2*X1*X2*Y2) / (2*(X1^2+2*X1*X2+X2^2+Y1^2-2*Y1*Y2+Y2^2));
L = sqrt(Xe^2 + Ye^2);
sinq = Xe / L;
cosq = Ye / L;

% 雅可比矩阵元素计算
q1_L = ((Xe+X1)*sinq + (Ye-Y1)*cosq)/(L1u*(-(Xe+X1)*sin(q1) + (Ye-Y1)*cos(q1)));
q1_q = (L*((Xe+X1)*cosq - (Ye-Y1)*sinq))/(L1u*(-(Xe+X1)*sin(q1) + (Ye-Y1)*cos(q1)));
q2_L = ((Xe-X2)*sinq + (Ye-Y2)*cosq)/(L2u*((Xe-X2)*sin(q2) + (Ye-Y2)*cos(q2)));
q2_q = (L*((Xe-X2)*cosq - (Ye-Y2)*sinq))/(L2u*((Xe-X2)*sin(q2) + (Ye-Y2)*cos(q2)));

%定义矩阵A的元素(对角线)

a = q1_L;
b = q1_q;
c = q2_L;
d = q2_q;


%计算行列式

detA = a*d - b*c;

%计算逆矩阵
J_inv = (1/detA) * [d -b;-c a];

%计算J逆的转置
J = J_inv';

F = [F0;Tp];
T = J * F;
T1 = T(1,1);
T2 = T(2,1);

end

File: src/modules/leg_wbr/leg_wbr.c
//
// Created by SuperChen on 2025/11/9.
//

#include "leg_wbr.h"
#include "arm_math.h"


#define LEFT 0
#define RIGHT 1
static uint8_t idx = 0; // register idx,是该文件的全局轮腿索引,在注册时使用
static wbr_leg_obj_t wbr_obj[2];




static void wbr_calc(struct wbr_leg_obj *leg,struct ins_msg *ins,float dt){

    leg->thetab = ins->pitch * DEGREE_2_RAD;//机体倾角,沿机体正方向前倾为正,注意正负号,实测需要修正

    leg->X1 = leg->L1a - leg->L1u * arm_cos_f32(leg->phi1);
    leg->X2 = leg->L2a - leg->L2u * arm_cos_f32(leg->phi2);
    leg->Y1 = leg->L1u * arm_sin_f32(leg->phi1);
    leg->Y2 = leg->L2u * arm_sin_f32(leg->phi2);
    leg->sigma1 = sqrt(-(-leg->L1d*leg->L1d-2*leg->L1d*leg->L2d-leg->L2d*leg->L2d+leg->X1*leg->X1+2*leg->X1*leg->X2+leg->X2*leg->X2+leg->Y1*leg->Y1-2*leg->Y1*leg->Y2+leg->Y2*leg->Y2) * (-leg->L1d*leg->L1d+2*leg->L1d*leg->L2d-leg->L2d*leg->L2d+leg->X1*leg->X1+2*leg->X1*leg->X2+leg->X2*leg->X2+leg->Y1*leg->Y1-2*leg->Y1*leg->Y2+leg->Y2*leg->Y2));
    leg->Xe = (leg->Y1*leg->sigma1-leg->Y2*leg->sigma1+leg->L1d*leg->L1d*leg->X1+leg->L1d*leg->L1d*leg->X2-leg->L2d*leg->L2d*leg->X1-leg->L2d*leg->L2d*leg->X2+leg->X1*leg->X2*leg->X2-leg->X1*leg->X1*leg->X2-leg->X1*leg->Y1*leg->Y1-leg->X1*leg->Y2*leg->Y2+leg->X2*leg->Y1*leg->Y1+leg->X2*leg->Y2*leg->Y2-leg->X1*leg->X1*leg->X1+leg->X2*leg->X2*leg->X2+2*leg->X1*leg->Y1*leg->Y2-2*leg->X2*leg->Y1*leg->Y2) / (2*(leg->X1*leg->X1+2*leg->X1*leg->X2+leg->X2*leg->X2+leg->Y1*leg->Y1-2*leg->Y1*leg->Y2+leg->Y2*leg->Y2));
    leg->Ye = (leg->X1*leg->sigma1+leg->X2*leg->sigma1-leg->L1d*leg->L1d*leg->Y1+leg->L1d*leg->L1d*leg->Y2+leg->L2d*leg->L2d*leg->Y1-leg->L2d*leg->L2d*leg->Y2+leg->X1*leg->X1*leg->Y1+leg->X1*leg->X1*leg->Y2+leg->X2*leg->X2*leg->Y1+leg->X2*leg->X2*leg->Y2-leg->Y1*leg->Y2*leg->Y2-leg->Y1*leg->Y1*leg->Y2+leg->Y1*leg->Y1*leg->Y1+leg->Y2*leg->Y2*leg->Y2+2*leg->X1*leg->X2*leg->Y1+2*leg->X1*leg->X2*leg->Y2) / (2*(leg->X1*leg->X1+2*leg->X1*leg->X2+leg->X2*leg->X2+leg->Y1*leg->Y1-2*leg->Y1*leg->Y2+leg->Y2*leg->Y2));
    leg->L = sqrt(leg->Xe*leg->Xe + leg->Ye*leg->Ye);
    leg->q = atan2f(leg->Xe / leg->L,leg->Ye / leg->L);

    leg->d_L = (leg->L - leg->last_L)/dt;
    leg->dd_L = (leg->d_L - leg->last_d_L)/dt;
    leg->d_q = (leg->q - leg->last_q)/dt;
    leg->wbr_theta = leg->q + leg->thetab;
    leg->wbr_d_theta = (leg->wbr_theta - leg->last_wbr_theta)/dt;
    leg->wbr_dd_theta = (leg->wbr_d_theta - leg->last_d_wbr_theta)/dt;

    leg->last_L = leg->L;
    leg->last_d_L = leg->d_L;
    leg->last_q = leg->q;
    leg->last_wbr_theta = leg->wbr_theta;
    leg->last_d_wbr_theta = leg->wbr_d_theta;

}

static void wbr_calculation_Torque(struct wbr_leg_obj *leg,float *WBR_T){

    //雅可比矩阵元素计算
    leg->q1_L = ((leg->Xe+leg->X1)*arm_sin_f32(leg->q) + (leg->Ye-leg->Y1)*arm_cos_f32(leg->q))/(leg->L1u*(-(leg->Xe+leg->X1)*arm_sin_f32(leg->phi1) + (leg->Ye-leg->Y1)*arm_cos_f32(leg->phi1)));
    leg->q1_q = (leg->L*((leg->Xe+leg->X1)*arm_cos_f32(leg->q) - (leg->Ye-leg->Y1)*arm_sin_f32(leg->q)))/(leg->L1u*(-(leg->Xe+leg->X1)*arm_sin_f32(leg->phi1) + (leg->Ye-leg->Y1)*arm_cos_f32(leg->phi1)));
    leg->q2_L = ((leg->Xe-leg->X2)*arm_sin_f32(leg->q) + (leg->Ye-leg->Y2)*arm_cos_f32(leg->q))/(leg->L2u*((leg->Xe-leg->X2)*arm_sin_f32(leg->phi2) + (leg->Ye-leg->Y2)*arm_cos_f32(leg->phi2)));
    leg->q2_q = (leg->L*((leg->Xe-leg->X2)*arm_cos_f32(leg->q) - (leg->Ye-leg->Y2)*arm_sin_f32(leg->q)))/(leg->L2u*((leg->Xe-leg->X2)*arm_sin_f32(leg->phi2) + (leg->Ye-leg->Y2)*arm_cos_f32(leg->phi2)));

    //定义矩阵A的元素(对角线)

    leg->a = leg->q1_L;
    leg->b = leg->q1_q;
    leg->c = leg->q2_L;
    leg->d = leg->q2_q;

    //计算行列式

    leg->detA = leg->a*leg->d - leg->b*leg->c;

    if(leg->detA != 0){
        leg->J_inv[0][0] = leg->d / leg->detA;
        leg->J_inv[0][1] = -leg->b / leg->detA;
        leg->J_inv[1][0] = -leg->c / leg->detA;
        leg->J_inv[1][1] = leg->a / leg->detA;

        leg->J_inv_T[0][0] = leg->J_inv[0][0];
        leg->J_inv_T[0][1] = leg->J_inv[1][0];
        leg->J_inv_T[1][0] = leg->J_inv[0][1];
        leg->J_inv_T[1][1] = leg->J_inv[1][1];

    }

    WBR_T[0] = leg->J_inv_T[0][0] * leg->wbr_Fbl + leg->J_inv_T[0][1] * leg->WBR_Tbl ;
    WBR_T[1] = leg->J_inv_T[1][0] * leg->wbr_Fbl + leg->J_inv_T[1][1] * leg->WBR_Tbl;
}

static int8_t input_wbr_leg_angle(struct wbr_leg_obj *leg, float phi1, float phi2){

    leg->phi1 = phi1;
    leg->phi2 = phi2;

    if(phi1 < leg->phi1_max && phi1 > leg->phi1_min){
        if(phi2 < leg->phi2_max && phi2 > leg->phi2_min){
            leg->wbr_leg_state = LEG_NORMAL;
        }else{
            leg->wbr_leg_state = LEG_ERROR;
        }
    }else{
        leg->wbr_leg_state = LEG_ERROR;
    }


    return (uint8_t)leg->wbr_leg_state;

}

/**
 * @brief 轮腿初始化,返回一个轮腿实例
 * @param config 轮腿配置
 * @return wbr_leg_obj_t* wbr建模轮腿实例指针
 */
wbr_leg_obj_t * wbr_leg_register(wbr_leg_config_t *config)
{
    wbr_obj[idx].L1u = config->L1u;
    wbr_obj[idx].L2u = config->L2u;
    wbr_obj[idx].L1d = config->L1d;
    wbr_obj[idx].L2d = config->L2d;
    wbr_obj[idx].L1a = config->L1a;
    wbr_obj[idx].L2a = config->L2a;
    wbr_obj[idx].phi1_max = PI;    //根据实际修改
    wbr_obj[idx].phi1_min = 0;
    wbr_obj[idx].phi2_max = PI;
    wbr_obj[idx].phi2_min = 0;
    wbr_obj[idx].wbr_leg_state = WBR_LEG_ERROR;
    wbr_obj[idx].q = PI/2;
    wbr_obj[idx].wbr_calc = wbr_calc;
    wbr_obj[idx].wbr_cal_T = wbr_calculation_Torque;
    wbr_obj[idx].input_wbr_leg_angle = input_wbr_leg_angle;

    return &wbr_obj[idx++];
}